### PR TITLE
403 betting game market wincase

### PR DESCRIPTION
**Pull request by Katkevich**

### CODE REVIEW

File: libraries/protocol/include/scorum/protocol/betting/wincase.hpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+{
+};
+
+template <market_kind kind, typename tag = void> struct under;
+template <market_kind kind, typename tag = void> struct over
+{
+    static constexpr market_kind kind_v = kind;
+    using opposite_type = under<kind, tag>;
+
+    float threshold;
```
Comment (alexplevako) [inline libraries/protocol/include/scorum/protocol/betting/wincase.hpp]: dont use float. create custom type.

File: libraries/protocol/include/scorum/protocol/betting/wincase.hpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        return opposite_type{ threshold };
+    }
+};
+
+template <market_kind kind, typename tag> struct under
+{
+    static constexpr market_kind kind_v = kind;
+    using opposite_type = over<kind, tag>;
+
+    threshold_type threshold = 0;
```
Comment (Andrei-Masilevich) [inline libraries/protocol/include/scorum/protocol/betting/wincase.hpp]: Not good architecture when fields are duplicated in initial and opposite wincase.
Maybe try additional bool tag. For instance:

```
template <bool site, market_kind kind, typename tag = void> struct over_under
{
    static constexpr market_kind kind_v = kind;
    using opposite_type = over_under<!site, kind, tag>;

    threshold_type threshold = 0;

    opposite_type create_opposite() const
    {
        return opposite_type{ threshold };
    }
};
template <market_kind kind, typename tag = void> using over = over_under<true, kind, tag>;
template <market_kind kind, typename tag = void> using under = over_under<false, kind, tag>;
```

File: libraries/protocol/include/scorum/protocol/betting/wincase_comparison.hpp [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+{
+    return lhs.threshold < rhs.threshold;
+}
+
+template <market_kind kind, typename tag>
+bool operator<(const under<kind, tag>& lhs, const under<kind, tag>& rhs)
+{
+    return lhs.threshold < rhs.threshold;
+}
+
```
Comment (Andrei-Masilevich) [inline libraries/protocol/include/scorum/protocol/betting/wincase_comparison.hpp]: Use one equivalent of your two functions:
```
template <bool site, market_kind kind, typename tag>
bool operator<(const over_under<site, kind, tag>& lhs, const over_under<site, kind, tag>& rhs)
{
    return lhs.threshold < rhs.threshold;
}
```

File: tests/utests/betting/invariants_validation_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    BOOST_REQUIRE_THROW(validate_markets(markets), fc::assert_exception);
+}
+
+SCORUM_TEST_CASE(validate_game_full_market_list_test_positive)
+{
+    soccer_game game;
+    fc::flat_set<market_type> markets
+        = { { market_kind::result, { { result_home{}, result_draw_away{} } } },
+            { market_kind::round, { { round_home{}, round_away{} } } },
+            { market_kind::handicap, { { handicap_home_over{ 500 }, handicap_home_over{ 500 } } } },
```
Comment (Andrei-Masilevich) [inline tests/utests/betting/invariants_validation_tests.cpp]: Where is handicap_home_under? And there is no throw...

File: tests/utests/betting/game_serialization_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+{
+    create_game_operation get_soccer_game_operation() const
+    {
+        create_game_operation op;
+        op.moderator = "moderator_name";
+        op.name = "game_name";
+        op.game = soccer_game{};
+        op.start = time_point_sec{ 1461605400 };
+        op.markets = { { market_kind::result,
+                         { { result_home{}, result_draw_away{} },
```
Comment (Andrei-Masilevich) [inline tests/utests/betting/game_serialization_tests.cpp]: Why do you duplicate wincases: initial - opposite. Is it really impossible to extract opposite wincase?
Maybe use over_under, yes_no instead

**Pull request merged**